---
title: "Tutorial from DGEA to motifs search"
output: html_document
---

In this RMarkdown document, you will find instructions to perform the differential gene expression analysis (DGEA), results of which are used for the enriched motif search in the promoters of the differentially expressed genes (DEGs). The motif search can be performed with the MEME suite in multiple modes <https://meme-suite.org/meme/index.html>. One mode is the de novo search with the MEME tool; search for the enriched known motifs is done with the AME tool. After the AME mnotif search is finished, you can use the code at the end of the document to perform the initial comparison of the AME results for the up and down regulated genes.

# Load data for DGEA and necessary packages

```{r input, echo=TRUE, message=FALSE, warning=FALSE}
# load the necessary libraries
library("tidyverse")
library("ggplot2")
library("UpSetR")

# read raw read counts and the design matrix
raw_counts <- read_tsv("../data/BB-RNAseq-input-datasets/PRJEB25079_RawCounts_reformatted.tsv")
design <- read_tsv("../data/BB-RNAseq-input-datasets/design-matrix.tsv")

```

# Initial exploration of the read count table

The read count table contains 319 samples (one column per sample), and the first column has the gene names. The sample IDs are in the column names. In total, the expression of 26,397 genes was evaluated in this RNAseq experiment (\# of rows).

```{r initial_count_exploration1, echo=TRUE, message=FALSE, warning=FALSE}
# show top rows to see the overall structure of the read count table
print("showing the top rows in the read count table")
head(raw_counts)

# check the total number of samples
print(paste("the total number of samples is", ncol(raw_counts)-1))

# check the total number of genes
print(paste("the total number of genes is", nrow(raw_counts)))

```

# DGEA in DESeq2

```{r DGEA_DESeq2, echo=TRUE, message=FALSE, warning=FALSE}
#BiocManager::install("DESeq2")
library("DESeq2")

# move the first column to rownames
raw_counts <- raw_counts %>% column_to_rownames(var = "GeneID")

### create a DESeq2 object
# since the samples differ by only one parameter, we need to select only this parameter from the design table, otherwise DESeq2 does not work
keep_factor <- c("Treatment")

# select samples for the analysis:
keep_sample <- c("Col_mock_030", "Col_mock_0301", "Col_mock_0302", "Col_mock_0303", "Col_flg22_030", "Col_flg22_0301", "Col_flg22_0302", "Col_flg22_0303")

# reformat the design table for DESeq2 (move sample names to row names) and subset the design table to contain only the experimental factors of interest
DESeq2_design <- design %>% column_to_rownames(var = "SampleID")
DESeq2_design <- DESeq2_design[keep_sample, ] %>% select(all_of(keep_factor))

print("create DESeq2 object")
dds <- DESeqDataSetFromMatrix(countData = raw_counts[, keep_sample],
                              colData = DESeq2_design,
                              design = ~ Treatment) # add Timepoint if you include samples from different timepoints too


# check the imported object
dim(dds) # dimensions
head(rownames(dds)) # gene names
head(colnames(dds)) # sample names

# discard genes that are not or low expressed; keep those that are expressed at least in two samples in the dataset
print(paste("Total number of genes:", nrow(dds)))
keep <- rowSums(counts(dds) >= 5) >= 3 #genes with more than 5 counts in more than 2 samples
dds <- dds[keep,]
print(paste("Number of genes after additional filtering:", nrow(dds)))

### data transformation for PCA
print("transform counts using the variance stabilizing transformation; one could also perform PCA on cpm-normalized data but the variance-stabilizing transformation is preferred")
transformed_counts <- vst(dds, blind = FALSE)

# PCA with color based on different parameters using argument "intgroup"
for (exp_factor in list(c("Treatment"))){
  pcaData <- plotPCA(transformed_counts, intgroup = exp_factor, ntop = nrow(transformed_counts), returnData = TRUE)
  percentVar <- round(100 * attr(pcaData, "percentVar"))
  g <- ggplot(pcaData, aes(x = PC1, y = PC2, color = factor(group)))
  g <- g + geom_point(size = 4, alpha = 1) +
    xlab(paste0("PC1: ", percentVar[1], "% variance")) +
    ylab(paste0("PC2: ", percentVar[2], "% variance")) +
    theme_classic()
}

# display the latest PCA plot in the RMarkdown report.
g

### fit the model for the differential gene expression (DGE) analysis
dds <- DESeq(dds, test = "Wald")

### extract results of the DGE analysis
res <- results(dds, contrast = c("Treatment", "flg22", "mock")) # base for the comparison at the end
overview_res <- data.frame(GeneID = row.names(res), log2FoldChange = res$log2FoldChange, p_adj = res$padj)

# thresholds
padj_cutoff <- 0.05
lfc_cutoff <- 1

# calculate the number of DEGs
up_deseq2 <- overview_res %>% filter(log2FoldChange >= lfc_cutoff & p_adj <= padj_cutoff) %>% select(GeneID)
down_deseq2 <- overview_res %>% filter(log2FoldChange <= -lfc_cutoff & p_adj <= padj_cutoff) %>% select(GeneID)

```

# Write lists of up and downregulated genes in separate files

```{r write_results_in_files, echo=TRUE, message=FALSE, warning=FALSE}
# make directory in the working directory to write the results there
dir.create("results-deseq2-for-MEME")

# write deseq2 results
write_tsv(up_deseq2, "./results-deseq2-for-MEME/up-genes.tsv")
write_tsv(down_deseq2, "./results-deseq2-for-MEME/down-genes.tsv")
```

# Load results of the AME search and compare them

```{r ame_results_filtering1, echo=TRUE, message=FALSE, warning=FALSE}
# read results of AME in the TSV format
up_ame <- read_tsv("./results-AME/up-AME-results.txt")
down_ame <- read_tsv("./results-AME/down-AME-results.txt")
up_deseq2 <- read_tsv("./results-deseq2-for-MEME/up-genes.tsv")
down_deseq2 <- read_tsv("./results-deseq2-for-MEME/down-genes.tsv")

# check format of the read files
head(up_ame)
head(down_ame)

# select only significant motifs
up_ame <- up_ame %>% filter(`adj_p-value` <= 0.05 & `E-value` <= 0.001) %>% select(motif_alt_ID) %>% pull()
down_ame <- down_ame %>% filter(`adj_p-value` <= 0.05 & `E-value` <= 0.001) %>% select(motif_alt_ID) %>% pull()

# define the sets of genes for comparison and combine them in one list
sets <-list(up_motifs = up_ame,
            down_motifs = down_ame)

# the list is then used as an input for the upset function
upset(fromList(sets), order.by = "freq")

```

We see a large set of the motifs specifically enriched among the promoters of the upregulated genes. We can find what these motifs are by running the code below. What is your interpretation of the result? Are the flg22 set-specific motifs also reported as enriched in Bjornson et al., 2021?

```{r ame_results_filtering2, echo=TRUE, message=FALSE, warning=FALSE}
# motifs specific to upregulated genes
print("motifs specific to upregulated genes")
setdiff(up_ame, down_ame)


print("motifs specific to downregulated genes")
setdiff(down_ame, up_ame)
```
